src/hotspot/share/jfr/recorder/service/jfrRecorderService.cpp

```diff
@@ -121,6 +122,30 @@ const Thread* JfrRotationLock::_owner_thread = nullptr;
 const int JfrRotationLock::retry_wait_millis = 10;
 volatile int JfrRotationLock::_lock = 0;
 
+// Reset thread local state used for object allocation sampling.
+class ClearObjectAllocationSampling : public ThreadClosure {
+ public:
+  void do_thread(Thread* t) {
+    assert(t != nullptr, "invariant");
+    t->jfr_thread_local()->clear_last_allocated_bytes();
+  }
+};
+
+template <typename Iterator>
+static inline void iterate(Iterator& it, ClearObjectAllocationSampling& coas) {
+  while (it.has_next()) {
+    coas.do_thread(it.next());
+  }
+}
+
+static void clear_object_allocation_sampling() {
+  ClearObjectAllocationSampling coas;
+  JfrJavaThreadIterator jit;
+  iterate(jit, coas);
+  JfrNonJavaThreadIterator njit;
+  iterate(njit, coas);
```

Comment:
Apologies for the post-integration review, but I have a question/suggestion. This code retains the JavaThread iterator while NonJavaThreadIterator is running, which feels unnecessary. Also, if we have thousands of threads, walking these threads and doing the unconditional `clear_last_allocated_bytes` there seems redundant and heavy-weight if the relevant JFR event is off.

Reply:
At this point, no recording is running, and no event settings are applied. No state exists to know if the event was enabled when the previous last recording stopped.

Reply:
OK, that's unfortunate. 

I would still suggest to split the scopes for Java and non-Java iterators. And maybe check if `last_allocated_bytes` is not zero (which is a rare case, especially when event is disabled) before clearing, thus replacing the unconditional write with conditional read-write.